idealista-scraper/idealista.py

```diff
@@ -0,0 +1,172 @@
+
+    # Features
+    data["features"] = {}
+    #  first we extract each feature block like "Basic Features" or "Amenities"
+    for feature_block in selector.css(".details-property-h2"):
+        # then for each block we extract all bullet points underneath them
+        label = feature_block.xpath("text()").get()
+        features = feature_block.xpath("following-sibling::div[1]//li")
+        data["features"][label] = ["".join(feat.xpath(".//text()").getall()).strip() for feat in features]
+
+    # Images
+    # the images are tucked away in a javascript variable.
+    # We can use regular expressions to find the variable and parse it as a dictionary:
+    image_data = re.findall("fullScreenGalleryPics\s*:\s*(\[.+?\]),", response.scrape_result["content"])[0]
+    # we also need to replace unquoted keys to quoted keys (i.e. title -> "title"):
+    images = json.loads(re.sub(r"(\w+?):([^/])", r'"\1":\2', image_data))
+    data["images"] = defaultdict(list)
+    data["plans"] = []
+    for image in images:
+        url = urljoin(str(response.context["url"]), image["imageUrl"])
+        if image["isPlan"]:
+            data["plans"].append(url)
+        else:
+            data["images"][image["tag"]].append(url)
+    return data
+
+
+async def scrape_provinces(urls: List[str]) -> List[str]:
+    """
+    Scrape province pages like:
+    https://www.idealista.com/en/venta-viviendas/balears-illes/con-chalets/municipios
+    for search page urls like:
+    https://www.idealista.com/en/venta-viviendas/marbella-malaga/con-chalets/
+    """
+    # add province pages to a scraping list
+    to_scrape = [ScrapeConfig(url, **BASE_CONFIG) for url in urls]
+    search_urls = []
+    # scrape the province pages concurrently
+    async for response in SCRAPFLY.concurrent_scrape(to_scrape):
+        search_urls.extend(parse_province(response))
+    log.success(f"scraped {len(search_urls)} search URLs")
+    return search_urls
+
+
+def parse_search(response: ScrapeApiResponse) -> List[str]:
+    """Parse search result page for 30 listings"""
+    selector = response.selector
+    urls = selector.css("article.item .item-link::attr(href)").getall()
+    return [urljoin(str(response.context["url"]), url) for url in urls]
+
+
+async def scrape_properties(urls: List[str]) -> List[PropertyResult]:
+    """Scrape Idealista.com properties"""
+    properties = []
+    # add all property pages to a scraping list
+    to_scrape = [ScrapeConfig(url, **BASE_CONFIG) for url in urls]
+    async for response in SCRAPFLY.concurrent_scrape(to_scrape):
+        # skip invalid property pages
+        if response.upstream_status_code != 200:
+            log.warning(f"can't scrape property: {response.context['url']}")
+            continue
+        properties.append(parse_property(response))
+    log.success(f"scraped {len(properties)} property listings")
+    return properties
+
+
+async def scrape_search(url: str, scrape_all_pages: bool, max_scrape_pages: int) -> List[str]:
```

Comment:
scrape_all_pages seems redundant here. If max_scrape_pages not set then we can imply all pages should be scraped.